iam_groups_authn/sync.py

```diff
@@ -48,142 +51,192 @@ async def groups_sync(iam_groups, sql_instances, credentials, private_ip=False):
-            users_with_roles_task = asyncio.create_task(
-                get_users_with_roles(role_service, role)
-            )
+            # run all the mapped syncs
+            results = await asyncio.gather(*sync_tasks, return_exceptions=True)
 
-            # await dependent tasks
-            results = await asyncio.gather(
-                add_users_task, verify_role_task, return_exceptions=True
-            )
-            # raise exception if found
+            # if one of the syncs fails, fail entire run
             for result in results:
                 if issubclass(type(result), Exception):
                     raise result
 
-            # log IAM users added as database users
-            added_users = results[0]
-            logging.debug(
-                "[%s][%s] Users added to database: %s."
-                % (instance, group, list(added_users))
-            )
 
-            # revoke group role from users no longer in IAM group
-            revoke_role_task = asyncio.create_task(
-                revoke_iam_group_role(
-                    role_service,
-                    role,
-                    users_with_roles_task,
-                    group_tasks[group],
-                    database_version,
-                )
+async def sync_group(
+    group,
+    instance,
+    group_task,
+    instance_tasks,
+    group_roles,
+    user_service,
+    credentials,
+    ip_type,
+):
+    """
+    Sync the IAM members of a single group to a single Cloud SQL instance.
+    """
+    try:
+        database_version = await instance_tasks[1]
+        # verify that group role for database won't exceed character limit
+        verify_group_role_length(group, group_roles, database_version)
+        # add missing IAM group members to database
+        add_users_task = asyncio.create_task(
+            add_missing_db_users(
+                user_service,
+                group_task,
+                instance_tasks[0],
+                instance,
+                database_version,
             )
+        )
 
-            # grant group role to IAM users who are missing it on database
-            grant_role_task = asyncio.create_task(
-                grant_iam_group_role(
-                    role_service,
-                    role,
-                    users_with_roles_task,
-                    group_tasks[group],
-                    database_version,
-                )
-            )
-            results = await asyncio.gather(
-                revoke_role_task, grant_role_task, return_exceptions=True
-            )
-            # raise exception if found
-            for result in results:
-                if issubclass(type(result), Exception):
-                    raise result
+        # initialize database connection pool
+        if database_version.is_mysql():
+            db = init_mysql_connection_engine(instance, credentials, ip_type)
+            role_service = MysqlRoleService(db)
+        else:
```

Comment:
nit: I know this is existing logic, but the following would be safer:
```
if db_v.is_mysql():
   # ...
else if db_v.is_postgres():
   # ...
else
    throws Unsupported DB error
```

Reply:
If the code gets to this point the database can only be postgres or mysql since the unsupported database error will have already been thrown. https://github.com/GoogleCloudPlatform/cloud-sql-iam-db-authn-groups/blob/map-group-names/iam_groups_authn/sync.py#L351

Reply:
Added suggestion with `elif` for postgres